client: redo reserves

chappjc opened the pull request.

martonp commented on client/core/reserves.go:
```diff
@@ -0,0 +1,171 @@
+	refunder.UnlockRefundReserves(unlock)
+}
+
+func (t *trackedTrade) unreserveMatches(count uint32) {
```
`unreserveLots`?

martonp commented on client/core/reserves.go:
```diff
@@ -0,0 +1,171 @@
+	total, err := redeemer.ReserveNRedemptions(uint64(count), t.wallets.toAsset)
+	if err != nil {
+		t.dc.log.Errorf("error reserving for %d redemptions: %v", count, err)
+		return // that's alright?
```
I think this error should be returned. This could mean that the wallet doesn't have enough to pay for the redemption tx fees.

martonp reviewed (commented):
I've just looked through this quickly, but it I like the changes. @buck54321 wrote the original code, so he should comment also.

chappjc commented:
Just keeping this draft for now.  If reserves are now working great we don't need the upheaval, although the "dust" heuristics need to go away at some point.

chappjc commented on client/core/core.go:
```diff
@@ -4521,19 +4534,19 @@ func (c *Core) prepareTrackedTrade(dc *dexConnection, form *TradeForm, crypter e
+			LotSize:          lotSize,
+			ChangeCoin:       changeID,
+			MaxFeeRate:       wallets.fromAsset.MaxFeeRate,
+			RedeemMaxFeeRate: wallets.toAsset.MaxFeeRate,
```
**Bug fix** to extract from this PR. (`RedeemMaxFeeRate` not stored)

chappjc commented on client/core/core.go:
```diff
@@ -5099,11 +5112,11 @@ func (c *Core) reFee(wallet *xcWallet, dc *dexConnection) {
 }
 
-func (c *Core) dbOrders(dc *dexConnection) ([]*db.MetaOrder, error) {
+func (c *Core) dbOrders(host string) ([]*db.MetaOrder, error) {
```
Possible simplification to extract from this PR.

chappjc commented on client/core/trade.go:
```diff
@@ -133,209 +132,72 @@ type trackedTrade struct {
+
+// newTrackedTrade is a constructor for a trackedTrade. The dexConnection must
+// have a valid config.
+func newTrackedTrade(dbOrder *db.MetaOrder, preImg order.Preimage, dc *dexConnection,
```
Possible simplification to extract from this PR. (`epochLen` is in the `dc`, so a redundant arg)

chappjc commented on client/core/trade.go:
```diff
@@ -1537,26 +1395,19 @@ func (t *trackedTrade) revoke() {
 // revokeMatch sets the status as revoked for the specified match. revokeMatch
 // must be called with the mtx write-locked.
-func (t *trackedTrade) revokeMatch(ctx context.Context, matchID order.MatchID, fromServer bool) error {
+func (t *trackedTrade) revokeMatch(matchID order.MatchID, fromServer bool) error {
```
Possible simplification to extract from this PR. (context arg unused)

chappjc commented on client/core/trade.go:
```diff
@@ -666,40 +527,26 @@ func (t *trackedTrade) negotiate(msgMatches []*msgjson.Match) error {
 	if cancelMatch != nil {
-		newFill += cancelMatch.Quantity
-	}
-	// The filled amount includes all of the trackedTrade's matches, so the
-	// filled amount must be set, not just increased.
-	trade.SetFill(filled)
```
Possible simplification to extract from this PR (`recalcFilled` does this now)

chappjc commented on client/core/trade.go:
```diff
@@ -744,6 +591,11 @@ func (t *trackedTrade) recalcFilled() (matchFilled, canceled uint64) {
 	if t.cancel != nil && t.cancel.matches.maker != nil {
 		canceled = t.cancel.matches.maker.Quantity
 	}
+	return
+}
+
+func (t *trackedTrade) recalcFilled() (matchFilled, canceled uint64) {
+	matchFilled, canceled = t.filled()
```
Possible refactor to extract from this PR. (get `matchFilled` amount without calling `SetFill`)

chappjc commented on client/core/trade.go:
```diff
@@ -958,11 +810,6 @@ func (t *trackedTrade) isActive() bool {
 
 	// Status of all matches for the order.
 	for _, match := range t.matches {
-		proof := &match.MetaData.Proof
-		t.dc.log.Tracef("Checking match %s (%v) in status %v. "+
-			"Order: %v, Refund coin: %v, ContractData: %x, Revoked: %v", match,
-			match.Side, match.Status, t.ID(),
-			proof.RefundCoin, proof.ContractData, proof.IsRevoked())
```
Change to extract from this PR.  (This TRC log was handy once when we were debugging order status issues, but now it is just noise)

chappjc commented on client/db/bolt/db.go:
```diff
@@ -699,18 +698,21 @@ func (db *BoltDB) UpdateOrder(m *dexdb.MetaOrder) error {
 			put(maxFeeRateKey, uint64Bytes(md.MaxFeeRate)).
 			put(redeemMaxFeeRateKey, uint64Bytes(md.RedeemMaxFeeRate)).
-			put(redemptionFeesKey, uint64Bytes(md.RedemptionFeesPaid)).
+			put(fromVersionKey, uint32Bytes(md.FromVersion)).
+			put(toVersionKey, uint32Bytes(md.ToVersion)).
```
**Bug fix** to extract from this PR. ( `UpdateOrder` was not storing `ToVersion` or `FromVersion`)

chappjc commented on client/db/bolt/db.go:
```diff
@@ -1131,13 +1127,8 @@ func (db *BoltDB) UpdateOrderMetaData(oid order.OrderID, md *db.OrderMetaData) e
 			put(changeKey, md.ChangeCoin).
 			put(linkedKey, linkedB).
 			put(swapFeesKey, uint64Bytes(md.SwapFeesPaid)).
-			put(maxFeeRateKey, uint64Bytes(md.MaxFeeRate)).
 			put(redemptionFeesKey, uint64Bytes(md.RedemptionFeesPaid)).
-			put(fromVersionKey, uint32Bytes(md.FromVersion)).
-			put(toVersionKey, uint32Bytes(md.ToVersion)).
 			put(optionsKey, config.Data(md.Options)).
-			put(redemptionReservesKey, uint64Bytes(md.RedemptionReserves)).
-			put(refundReservesKey, uint64Bytes(md.RefundReserves)).
```
Possible simplification to extract from this PR. (`UpdateOrderMetaData` should not update these constant values that are supposed to be stored on creation in `UpdateOrder`)

chappjc reviewed (commented):
Still keeping on the back burner, but this review notes bugs and other changes that we should pull out of this and get merged separately.

The failures to store certain data in the DB may be critical.

chappjc commented:
Rebased and updated for live reconfigure changes, but literally never tested so far.  Just making it closer to ready for https://github.com/decred/dcrdex/issues/1544
